### PR TITLE
CUMULUS-3449:Update cumulus_id type add indexes

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-3449: SPIKE: Improve performance for reconciliation query](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3449)

## Changes

* Update Cumulus_id Type and Indexes
* Add migration instruction

## Test Instructions:

1. Build and deploy your stack from master branch.
2. Add a few granules with modified spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js.  Comment out afterAll and skip the test steps:
xdescribe('when moving a granule'
xit('can delete the ingested granule from the API'
3. Follow migration instruction to update cumulus_id type and indexes. Verify that type and indexes have been updated.
4. Build from PR branch.  Update example/data-persistence-tf/main.tf, set dbRecreation=false.  Deploy data-persistence-tf module. Verify that granules_collection_cumulus_id_granule_id_unique UNIQUE CONSTRAINT is added and old constraint is dropped.

## PR Checklist

- [x] Update CHANGELOG
- [ ] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/db/src/migrations/20240125171703_update_granule_execution_cumulus_id_type.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { Knex } from 'knex';
+
+export const up = async (knex: Knex): Promise<void> => {
+  await knex.raw('ALTER TABLE files ALTER COLUMN granule_cumulus_id TYPE BIGINT');
```
Comment (jennyhliu) [inline packages/db/src/migrations/20240125171703_update_granule_execution_cumulus_id_type.ts:4]: type mismatch can cause files_granule_cumulus_id_index not being used.

File: packages/db/src/migrations/20240126135619_granules_add_indexes.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { Knex } from 'knex';
+
+export const up = async (knex: Knex): Promise<void> => {
+  await knex.raw('CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS granules_collection_cumulus_id_granule_id_unique ON granules(collection_cumulus_id, granule_id)');
```
Comment (Jkovarik) [inline packages/db/src/migrations/20240126135619_granules_add_indexes.ts:4]: How long does this take to run on a production database?

**Review by Jkovarik — CHANGES_REQUESTED** [review]
I realize this is in progress, but putting a blocking concern on addressing migration time scale issues.

File: packages/db/tests/lib/test-QuerySearchClient.js [hunk 3]
```diff
@@ -96,15 +96,15 @@ test('QuerySearchClient.shift() returns next record from current set of results
     await querySearchClient.shift(),
     {
       ...records[0],
-      granule_cumulus_id: Number.parseInt(records[0].granule_cumulus_id, 10),
```
Comment (charleshuang80) [inline packages/db/tests/lib/test-QuerySearchClient.js:99]: I'm curious why the change to BigInt would make the need for Number and parseInt to go away - does the change from Int to BigInt mean the value is being grabbed as a number now instead of a string?

Comment (jennyhliu) [reply packages/db/tests/lib/test-QuerySearchClient.js:99]: Bigint column returns as a string from knex query. files.granule_cumulus_id now is a bigint, not a int anymore.

Comment (charleshuang80) [reply packages/db/tests/lib/test-QuerySearchClient.js:99]: ok. sorry for my confusion, but so now granule_cumulus_id will be a string and granule_id will also be a string, yes?

Comment (jennyhliu) [reply packages/db/tests/lib/test-QuerySearchClient.js:99]: granules.granule_id is always a string.
files.granule_cumulus_id is changed from int to bigint which matches the granule.cumulus_id, and knex returns bigint as string.
`granule_cumulus_id: records[0].granule_cumulus_id` are both string now.

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  - Add indexes `granules_granule_id_index` and `granules_provider_collection_cumulus_id_granule_id_index`
+    to `granules` table
+
+The updates will be automatically created as part of the bootstrap lambda function on deployment of the data-persistence module.
+
+*In cases where the type update and indexes are already applied, the updates will have no effect.*
+
+## Apply the Changes in Production Environment
+
+In production and some UAT environments, the type updates and indexes must be applied manually since the commands can take significant
```
Comment (Jkovarik) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:37]: Suggest we make the distinction between 'large' and 'production'.    Users are going to want to know what row count (approximately) given DB size they can run the bootstrap versus this sort of apply, we should make it very clear what the difference is, or suggest any non-new deployment should use this procedure instead.

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+| Table Name | Original Table Size | ALTER Run Time (clock) | New Table Size | Number of Rows |
+|---|---|---|---|---|
+| executions | 1.8 TB | 41 hours | 1.4 TB | 374,170,855 |
+| files | 1.2 TB | 7.5 hours | 850 GB | 1,129,847,659 |
+| granule_executions | 20 GB | 1 hours | 13 GB | 182,566,709 |
+
+## Tools Used
+
+Since the update commands can take several hours to run based on table size and IO throughput, it is recommended that the commands are run in an EC2 instance
```
Comment (Jkovarik) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:51]: What happens if the query fails?

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+*In cases where the type update and indexes are already applied, the updates will have no effect.*
+
+## Apply the Changes in Production Environment
+
+In production and some UAT environments, the type updates and indexes must be applied manually since the commands can take significant
+amount of time. Since these particular ALTER TABLE commands require an EXCLUSIVE LOCK on the table, it is recommended that
+all database activity be quiesced. This means that Ingest and Archive and other Cumulus functions must be shutdown before and during these commands.
+
+The table below from LPDAAC provides the table sizes before and after `alter table` commands, and timings.
```
Comment (Jkovarik) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:41]: Nit: we should write this assuming readers don't know who LPDAAC is (target audience isn't only the I&A train)

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    as a secondary method of evaluating the database system state:
+
+    ```text
+    select pid, query, state, wait_event_type, wait_event from pg_stat_activity where state = 'active';
+    ```
+
+    If query rows are returned with a `query` value that involves the tables, make sure ingest is halted
+    and no other granule-update activity is running on the system.
+
+    Note: In rare instances if there are hung queries that are unable to resolve, it may be necessary to
```
Comment (Jkovarik) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:115]: If that occurs, what should users do after they ended the query?

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+There are many available resources for tmux and the psql client. The following is for your referrence.
+
+- [PostgreSQL Documentation](https://www.postgresql.org/docs/13/app-psql.html)
+- [Tmux tutorial](https://www.linuxtrainingacademy.com/tmux-tutorial/)
+
+:::note
+The version 9.x psql will be installed. This will work for all our standard SQL but some meta-commands like \d may not work.
+:::
+
+Basic commands for running SQL commands
```
Comment (Jkovarik) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:65]: Do these instructions (tmux commands in particular) work from an AWS console, or should users *only* use a console ssm session?

**Review by Jkovarik — COMMENTED** [review]
@jennyhliu  still reviewing, writing out initial comments from the review buffer, thanks for your patience.

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    tmux new-session -s CumulusUpgrade -n Fix-DataTypes
+
+    psql -h <Endpoint for writer instance> -p <Port for database or 5432> -d <cumulus database name> -U <database admin user> -W
+    #e.g. psql -h cumulus-dev-rds-cluster.cluster-xxx.us-east-1.rds.amazonaws.com -p 5432 -d cumulus_test_db -U cumulus_test -W
+
+    # Update column types
+    ALTER TABLE executions ALTER COLUMN cumulus_id TYPE BIGINT, ALTER COLUMN parent_cumulus_id TYPE BIGINT;
+    ALTER TABLE files ALTER COLUMN granule_cumulus_id TYPE BIGINT;
+    ALTER TABLE granules_executions ALTER COLUMN granule_cumulus_id TYPE BIGINT, ALTER COLUMN execution_cumulus_id TYPE BIGINT;
+    ALTER TABLE pdrs ALTER COLUMN execution_cumulus_id TYPE BIGINT;
```
Comment (charleshuang80) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:150]: So pdrs cumulus_id will never be big enough or run into the type mismatch? (I don't know much about pdrs so just wondering)

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    ```sh
+    sudo yum install -y tmux postgresql.x86_64
+    ```
+
+    Once installed, a tmux session is started with two windows. The Cumulus database is connected to in each window
+    using the PostgreSQL client. The primary window is used for running the ALTER commands, while the secondary window
+    is used to monitor the database and alter statement. When the operator hits end of shift or is done monitoring for
+    the day, the tmux session can be detached from and reattached to at a later time.
+
+4. Run SQL Commands
```
Comment (charleshuang80) [inline docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:137]: I think it would be helpful to add a line here that you should look at step 5 to monitor some of these commands. If I'm following the instructions line by line and am unfamiliar with tmux, which was the case for me, I might not see those instructions until I'm finished or almost finished with step 4.

**Review by charleshuang80 — COMMENTED** [review]
I was able to run all the migration instructions on my stack db and saw that they worked. Given that, and the tests all pass in Bamboo, I'm good with this PR. Have a few minor comments. But looks like Jonathan is reviewing so I'll hold off approving until he is good as well.

File: docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md (hunk 9 above)
Comment (jennyhliu) [reply docs/upgrade-notes/update-cumulus_id-type-indexes-CUMULUS-3449.md:150]: Most DAACs don't use PDRs, and one PDR can have multiple granules.

**Comment by Jkovarik** [comment]
@jennyhliu  the typing in the DB package for the modified fields are now incorrect (and the original typings may also be incorrect).  We need to update the modified typings to match the actual expected return from knex, and resolve all the downstream issues that may highlight.

Example return from a file.get query: 


```
bucket:
'80d'
checksum_type:
null
checksum_value:
null
created_at:
Fri Mar 01 2024 07:51:51 GMT-0700 (Mountain Standard Time)
cumulus_id:
'1'
file_name:
null
file_size:
null
granule_cumulus_id:
'1'
key:
'154'
path:
null
source:
null
type:
null
updated_at:
Fri Mar 01 2024 07:51:51 GMT-0700 (Mountain Standard Time)
[[Prototype]]:
Object
```

However the type for that record is now incorrect:

```
export interface PostgresFile {
  bucket?: string,
  key?: string,
  granule_cumulus_id: number,
  checksum_type?: string,
  checksum_value?: string,
  file_name?: string,
  file_size?: number,
  path?: string,
  source?: string,
  type?: string,
}
```

**Review by Jkovarik — CHANGES_REQUESTED** [review]
We need to resolve the mismatched types in the db package and downstream concerns

**Comment by jennyhliu** [comment]
@Jkovarik So the type for `cumulus_id` for granule and file are already wrong (number) in the master branch.  We can't just update `granule_cumulus_id` to string.  It's a big effort to update all related `cumulus_id` to be `string`.
> cumulus_id:
> '1'

**Review by Jkovarik — APPROVED** [review]
@jennyhliu  thanks for taking the time to address the typing feedback, this looks good!
